Add a method to return the keys in a section

hhschu opened the pull request:
Fix #6

I created a `.Keys()` method to return all the keys in a given section.

To achieve this, I also modified the structure of `_values`, to become `{"section1": {"key1": "value1", ...}, ...}`, and done a bit refactoring to increase readability and performance. We have a new method `.Get(section)` as a bonus.

SSARCandy commented on ini/INIReader.h:
```diff
@@ -263,13 +266,22 @@ class INIReader
-    std::map<std::string, std::string> _values;
-    std::set<std::string> _sections;
-    static std::string MakeKey(std::string section, std::string name);
+    std::map<std::string, std::map<std::string, std::string>> _values;
```
use `unordered_map` which is faster in this use case, ref: https://www.geeksforgeeks.org/unordered_map-in-cpp-stl/
> Internally unordered_map is implemented using Hash Table, the key provided to map are hashed into indices of hash table that is why performance of data structure depends on hash function a lot but on an average the cost of search, insert and delete from hash table is O(1).

SSARCandy replied:
or `unordered_set`

SSARCandy commented on ini/INIReader.h:
```diff
@@ -263,13 +266,22 @@ class INIReader
 
+    // Return the key-value pairs in the given section.
+    // Throw an error if scetion doesn't exist/
+    const _Get(std::string section) const;
```
should have return type

SSARCandy commented on ini/INIReader.h:
```diff
@@ -263,13 +266,22 @@ class INIReader
+
+    // Return the raw string values of a key under a given section.
+    // Return an empty string if the key doesn't exist.
+    const _Get(std::string section, std::string name) const;
```
should have return type

SSARCandy commented on ini/INIReader.h:
```diff
@@ -312,43 +324,38 @@ inline int INIReader::ParseError() const
+    return retval;
+}
+
+inline const std::set<std::string>& INIReader::Keys(std::string section) const
```
```
inline const std::set<std::string>& INIReader::Keys(std::string section) const
                                  ^
              this means return a reference.
```

In the above code, return value is a reference to the local variable `_setction `, which life time is limited in the function, so is invalid.

ref: https://stackoverflow.com/questions/4643713/c-returning-reference-to-local-variable

SSARCandy commented on ini/INIReader.h:
```diff
@@ -312,43 +324,38 @@ inline int INIReader::ParseError() const
 
 inline const std::set<std::string>& INIReader::Sections() const
```
same issue as `INIReader::Keys()`

SSARCandy commented on ini/INIReader.h:
```diff
@@ -380,9 +387,23 @@ inline std::vector<T> INIReader::GetVector(std::string section, std::string name
+inline const INIReader::_Get(std::string section, std::string name) const {
+    std::map<std::string, std::string> _section = _Get(section)
+    if (!_section.count(name)) {
+        return "";
```
why not throw an exception?

hhschu replied:
Because `.Get()` with a default value needs to catch the exception this way. I read that catching exception is expensive in C++, and I didn't see a "key not found" exception so it risks catching unexpected errors.

SSARCandy replied:
suggest implementataion, let `map::at()` handle access value or throw `out_of_range` error. ref: http://www.cplusplus.com/reference/map/map/at/

```cpp
inline const INIReader::_Get(std::string section, std::string name) const {
    return _section = _Get(section).at(name);
}
```

in my opinion, as a library, we need to throw error explicitly, not silently ignore or return special behavior (like returning an empty string), it may confuse ppl.

SSARCandy replied:
after read more carefully, i now understand you want `_Get` as an internal helper function, I would suggest you implement like this:

```cpp
template<typename T = std::string>
inline T INIReader::Get(std::string section, std::string name) const {
    try{
        std::string value = _values.at(section).at(name);
    } catch(std::out_of_range &e) { 
        // only handle key not found error, others err with throw directly.
        throw std::runtime_error("key " + section + name + " not found.");
    }
    return Parser<T>(value);
}
```

hhschu replied:
This doesn't work because `value` is in `try`'s scope.

SSARCandy reviewed (changes requested):
Please kindly read the comments and change them if you agree.
In my opinion, I think resolving the compile issue first would be easier for further development.

Commit:
fix: add boost library

Commit:
test: fix exit 1 when test failed

Commit:
test: fix get_vector_with_default

Test case 3 only has 2 elements. The comparison design requires all test cases
to have the same length as the first case.

Commit:
test: add test for .Sections() and .Keys()

Commit:
refactor: restructure _values

Changed `_values`'s structure to unordered_map to store the config
as {"section1": {"key1": "value1", ...}, ...}.

Also created a get section method and modified the way of checking
if a key exists. Previously, methods with a default value checks key
existence twice. Now all methods check and access a key only once.

Commit:
feat: add .Keys()

Add a new method to retrieve all keys in a given section.

Commit:
refactor: extract boolean conversion from .Get()

Extract boolean conversion to its own method and modified Converter
to increase readability of the constexpr logic in .Get().

SSARCandy commented on ini/INIReader.h:
```diff
@@ -310,57 +315,69 @@ inline int INIReader::ParseError() const
-template<typename T = std::string>
+inline const std::set<std::string> INIReader::Keys(std::string section) const
+{
+    auto _section = Get(section);
```
const

SSARCandy commented on ini/INIReader.h:
```diff
@@ -310,57 +315,69 @@ inline int INIReader::ParseError() const
-    if (!_values.count(key)) {
-        throw std::runtime_error("key " + key + " not found.");
+    auto _section = Get(section);
+    auto _value = _section.find(name);
```
const

SSARCandy reviewed (changes requested):
please add `const` qualifier if we do not intend to change the content.
And maybe we can try return reference in ` std::unordered_map<std::string, std::string> Get(std::string section) const` since we use it multiple times, returning reference would avoid calling copy constructor.

hhschu commented:
I thought about returning a reference. But in the end, I chose to return a value to be consistent with other methods. I also return a set in both `.Sections()` and `.Keys()` to be consistent with your previous interface, although I think it is more appropriate to return an array.

If you prefer to return a ref, I can change it.

SSARCandy commented on ini/INIReader.h:
```diff
@@ -370,48 +387,51 @@ inline std::vector<T> INIReader::GetVector(std::string section, std::string name
+        {"1", true}, {"true", true}, {"yes", true}, {"on", true},
+        {"0", false}, {"false", false}, {"no", false}, {"off", false},
+    };
+    auto value = s2b.find(s);
```
const

SSARCandy reviewed (approved):
LGTM 💯
